pkg/plugin/state.go

```diff
@@ -569,98 +569,205 @@ func (e *AutoscaleEnforcer) handleNodeDeletion(logger *zap.Logger, nodeName stri
+		cpuState = podResourceState[vmapi.MilliCPU]{
+			Reserved:         add.VCPU,
+			Buffer:           0,
+			CapacityPressure: 0,
+			Min:              add.VCPU,
+			Max:              add.VCPU,
+		}
+		memState = podResourceState[api.Bytes]{
+			Reserved:         add.Mem,
+			Buffer:           0,
+			CapacityPressure: 0,
+			Min:              add.Mem,
+			Max:              add.Mem,
+		}
 	}
 
+	podName := util.GetNamespacedName(pod)
+
+	ps := &podState{
+		name: podName,
+		node: node,
+		cpu:  cpuState,
+		mem:  memState,
+		vm:   vmState,
+	}
 	newNodeReservedCPU := node.cpu.Reserved + ps.cpu.Reserved
 	newNodeReservedMem := node.mem.Reserved + ps.mem.Reserved
 
-	cpuVerdict := fmt.Sprintf(
-		"node reserved %v + %v -> %v of total %v",
-		node.cpu.Reserved, ps.cpu.Reserved, newNodeReservedCPU, node.cpu.Total,
-	)
-	memVerdict := fmt.Sprintf(
-		"node reserved %v + %v -> %v of total %v",
-		node.mem.Reserved, ps.mem.Reserved, newNodeReservedMem, node.mem.Total,
-	)
+	verdict := verdictSet{
+		cpu: fmt.Sprintf(
+			"node reserved %v + %v -> %v of total %v",
+			node.cpu.Reserved, ps.cpu.Reserved, newNodeReservedCPU, node.cpu.Total,
+		),
+		mem: fmt.Sprintf(
+			"node reserved %v + %v -> %v of total %v",
+			node.mem.Reserved, ps.mem.Reserved, newNodeReservedMem, node.mem.Total,
+		),
+	}
 
-	log := logger.Info
-	if node.cpu.Reserved > node.cpu.Total || node.mem.Reserved > node.mem.Total {
-		log = logger.Warn
+	if allowDeny {
+		logger.Info("Allowing reserve resources for Pod", zap.Object("verdict", verdict))
+	} else if shouldDeny /* but couldn't */ {
+		logger.Warn("Reserved resources for Pod above totals", zap.Object("verdict", verdict))
+	} else {
+		logger.Info("Reserved resources for Pod", zap.Object("verdict", verdict))
 	}
 
-	log(
-		"Handled new non-VM pod",
-		zap.Object("verdict", verdictSet{
-			cpu: cpuVerdict,
-			mem: memVerdict,
-		}),
-	)
+	node.cpu.Reserved = newNodeReservedCPU
+	node.mem.Reserved = newNodeReservedMem
+
+	node.pods[podName] = ps
+	e.state.pods[podName] = ps
 
 	node.updateMetrics(e.metrics)
+
+	return true, &verdict, nil
 }
 
 // This method is /basically/ the same as e.Unreserve, but the API is different and it has different
 // logs, so IMO it's worthwhile to have this separate.
-func (e *AutoscaleEnforcer) handleVMDeletion(logger *zap.Logger, podName util.NamespacedName) {
+func (e *AutoscaleEnforcer) handleDeletion(logger *zap.Logger, podName util.NamespacedName) {
 	logger = logger.With(
 		zap.String("action", "VM deletion"),
 		zap.Object("pod", podName),
 	)
 
 	logger.Info("Handling deletion of VM pod")
 
+	logFields, kind, migrating, verdict := e.unreserveResources(logger, podName)
+
+	logger.With(logFields...).Info(
+		fmt.Sprintf("Deleted %s Pod", kind),
+		zap.Bool("migrating", migrating),
+		zap.Object("verdict", verdict),
+	)
+}
+
+func (e *AutoscaleEnforcer) unreserveResources(
+	logger *zap.Logger,
+	podName util.NamespacedName,
+) (_ []zap.Field, kind string, migrating bool, _ verdictSet) {
```

Comment:
To me having partially named return arguments looks slightly ugly. I'd rather suggest naming all or none.

Reply:
Returning log fields here looks sorta hacky.

Reply:
I'd left a comment on the function. I agree parts of this are a little ugly/hacky. I don't see a cleaner solution tbh, although maybe I'm missing something.

wdyt? 

(In terms of named/unnamed returns, IMO having a mix allows documenting some fields while making it clear that you shouldn't use a naked return. Here, I think it's necessary to name `kind` and `migrating`.)

Reply:
Sounds good.